python/ray/tests/test_client_builder.py

```diff
@@ -105,100 +105,114 @@ def ping(self):
+info = ray.client({address}).connect()
+print("NODE_ID:", ray.get_runtime_context().get_node_id())
 
-@ray.remote
-class Foo:
-    def ping(self):
-        return "pong"
+# Block.
+while True:
+    time.sleep(1)
+"""
 
-a = Foo.options(name="abc", lifetime="detached").remote()
-ray.get(a.ping.remote())
+    def _get_node_id(p: subprocess.Popen) -> str:
+        l = p.stdout.readline().decode("ascii").strip()
+        assert "NODE_ID" in l
+        return l[len("NODE_ID: ") :]
 
-import time
-while True:
-    time.sleep(30)
+    p1, p2, p3 = None, None, None
+    unbuffered = {"PYTHONUNBUFFERED": "1"}
+    try:
+        # ray.client() should start a cluster if none is running.
+        p1 = run_string_as_driver_nonblocking(
+            driver_template.format(address=""), env=unbuffered
+        )
+        p1_node_id = _get_node_id(p1)
 
-"""
-    blocking_local_script = driver_template.format(address="'local'", blocking=True)
-    blocking_noaddr_script = driver_template.format(address="", blocking=True)
-
-    # This should start a cluster.
-    p1 = run_string_as_driver_nonblocking(blocking_local_script)
-    # ray.client("local").connect() should start a second cluster.
-    p2 = run_string_as_driver_nonblocking(blocking_local_script)
-    # ray.client().connect() shouldn't connect to a cluster started by
-    # ray.client("local").connect() so it should create a third one.
-    p3 = run_string_as_driver_nonblocking(blocking_noaddr_script)
-    # ray.client().connect() shouldn't connect to a cluster started by
-    # ray.client().connect() so it should create a fourth one.
-    p4 = run_string_as_driver_nonblocking(blocking_noaddr_script)
-
-    wait_for_condition(
-        lambda: len(ray._private.services.find_gcs_addresses()) == 4,
-        retry_interval_ms=1000,
-    )
+        # ray.client("local") should always start a cluster.
+        p2 = run_string_as_driver_nonblocking(driver_template.format(address="'local'"))
+        p2_node_id = _get_node_id(p2)
 
-    p1.kill()
-    p2.kill()
-    p3.kill()
-    p4.kill()
-    # Prevent flakiness since fatesharing takes some time.
-    subprocess.check_output("ray stop --force", shell=True)
+        # ray.client() shouldn't connect to a cluster started by ray.client() or
+        # ray.client("local").
+        p3 = run_string_as_driver_nonblocking(driver_template.format(address=""))
+        p3_node_id = _get_node_id(p3)
 
-    # Since there's a cluster started with `ray start --head`
-    # we should connect to it instead.
-    subprocess.check_output("ray start --head", shell=True)
-    # The assertion in the driver should cause the script to fail if we start
-    # a new cluster instead of connecting.
-    run_string_as_driver(
-        """
-import ray
-ray.client().connect()
-assert len(ray._private.services.find_gcs_addresses()) == 1
+        # Check that all three drivers started their own local clusters.
+        assert len({p1_node_id, p2_node_id, p3_node_id}) == 3
```

Comment:
I know checking node id is better but do you know why `ray._private.services.find_gcs_addresses()` doesn't work?